tests/testsuite/cfg_features.rs

```diff
@@ -0,0 +1,363 @@
+            
+            pub fn bb() { let _ = BB; }
+        "#,
+        )
+        .build();
+    p.cargo("build -v")
+        .with_status(101)
+        .with_stderr_contains(
+            "\
+             error[E0425]: cannot find value `BB` in this scope",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn transitive() {
+    #[cfg(target_os = "macos")]
+    let config = "target_os = \"macos\"";
+    #[cfg(target_os = "windows")]
+    let config = "target_os = \"windows\"";
+    #[cfg(all(not(target_os = "macos"), not(target_os = "windows")))]
+    let config = "unix";
+
+    let p = project()
+        .no_manifest()
+        // root depends on a and c="1.1.0"
+        .file(
+            "root/Cargo.toml",
+            r#"
+            [package]
+			name = "root"
+			version = "0.0.1"
+			authors = []
+			
+			[dependencies]
+			a = { version = "*", path = "../a" }
+			c = { version = "1.1.0", path = "../c1" }
+        "#,
+        )
+        .file(
+            "root/src/main.rs",
+            r#"
+            fn main() {
+			    println!("Hello, world!");
+			}
+        "#,
+        )
+        // a depends on b and on OSX depends on b's flag maybe
+        .file(
+            "a/Cargo.toml",
+            &format!(
+                r#"
+				[package]
+				name = "a"
+				version = "0.1.0"
+				
+				[lib]
+				name = "a"
+				
+				[target.'cfg(not({}))'.dependencies]
+				b = {{ version = "*", path = "../b" }}
+				
+				[target.'cfg({})'.dependencies]
+				b = {{ version = "*", path = "../b", features = ["maybe"] }}
+		        "#,
+                config, config,
+            ),
+        )
+        .file(
+            "a/src/lib.rs",
+            r#"
+        "#,
+        )
+        // b depends on c="=1.0.0" if maybe is active.
+        .file(
+            "b/Cargo.toml",
+            r#"
+			[package]
+			name = "b"
+			version = "0.1.0"
+			
+			[dependencies]
+			c = { version = "1.0.0", path = "../c0", optional = true }
+			
+			[features]
+			maybe = ["c"]
+		"#,
+        )
+        .file(
+            "b/src/lib.rs",
+            r#"
+			#[cfg(feature = "maybe")]
+			pub fn maybe() {
+				c::cee();
+			}
+        "#,
+        )
+        // c 1.0.0
+        .file(
+            "c0/Cargo.toml",
+            r#"
+			[package]
+			name = "c"
+			version = "1.0.0"
+			
+			[lib]
+			name = "c"
+			
+			[dependencies]
+        "#,
+        )
+        .file(
+            "c0/src/lib.rs",
+            r#"			
+			pub fn cee() {}
+        "#,
+        )
+        // c 1.1.0
+        .file(
+            "c1/Cargo.toml",
+            r#"
+			[package]
+			name = "c"
+			version = "1.1.0"
+			
+			[lib]
+			name = "c"
+			
+			[dependencies]
+        "#,
+        )
+        .file(
+            "c1/src/lib.rs",
+            r#"
+        "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .cwd("root")
+        .with_stderr(
+            "\
+[COMPILING] c v1.0.0 ([..])
+[COMPILING] c v1.1.0 ([..])
```

Comment:
Sorry for the long delay in replying, I did not have internet access for the past week. This is passing because it uses both `c v1.0.0` and `c v1.1.0` witch is only allowed as they are path dependencies. What happens when using [`Package::new("c", "1.0.0").publish()`](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/generate_lockfile.rs#L69) to test as Crates.io dependencies?

(btw you can rebase or merge in the same branch instead of opening new prs)

Reply:
No worries. It may wait.
Have struggled fixing the windows' crlfs in place, so eventually had to recreate a PR.